File: ruby-rails/app/models/link.rb
class Link < ActiveRecord::Base
  attr_accessible :link_group_id, :title, :href, :description

  def self.create(user_id, attrs)
    model = Link.new attrs

    Link.do_access_check user_id, model.link_group_id do
      model.save

      return model
    end
  end

  def self.find_by_link_group(user_id, group_id)
    Link.do_access_check user_id, group_id do
      return Link.find_all_by_link_group_id group_id
    end
  end

  def self.delete_link(user_id, link_id)
    link = Link.find link_id

    Link.do_access_check user_id, link.link_group_id do
      link.delete

      return true
    end
  end

  def self.do_access_check(user_id, group_id)
    group = LinkGroup.where :id => group_id, :user_account_id => user_id

    if group.size() > 0
      yield
    else
      return nil
    end
  end
end

File: node-express/node/endpoints/link.js
var baseEndpoint = require('./base.js'),
    Link = require('../data/model/link.js'),
    LinkGroup = require('../data/model/linkGroup.js'),
    linkValidator = require('./validators/link-validator.js');

function getMethod(req, res, next)
{
    doAccessCheck(res, req.query.groupId, req.user.id, function() {
        res.format({
            'application/json': function() {
                Link.findByGroup(req.query.groupId, function(result) {
                    res.send(result);
                });
            }
        });
    });
}

function postMethod(req, res, next)
{
    var valid = linkValidator.validate(res, req.body);

    if (valid) {
        doAccessCheck(res, req.body.group.id, req.user.id, function() {
            res.format({
                'application/json': function() {
                    Link.save(req.body, function(result) {
                        res.send(201, result);
                    });
                }
            });
        });
    }
}

function deleteMethod(req, res, next)
{
    Link.findById(req.query.id, function(link) {
        if (link) {
            doAccessCheck(res, link.link_group_id, req.user.id, function() {
                res.format({
                    'application/json': function() {
                        Link.delete(req.query.id, function(result) {
                            res.send(result);
                        });
                    }
                });
            });
        }
        else {
            res.send(404, 'Link with id' + req.query.id + ' was not found.');
        }
    });
}

exports.create = baseEndpoint.createEndpoint({methodGet: getMethod, methodPost: postMethod, methodDelete: deleteMethod });

function doAccessCheck(res, groupId, userId, successCallback)
{
    LinkGroup.findByIdAndUser(groupId, userId, function(result) {
        if (result) {
            successCallback();
        }
        else {
            res.writeHead(409);
            res.end('The link group that was contained in the request is not valid for the current logged in user.');
        }
    });
}